protocol/binary/reader.go

```diff
@@ -21,465 +21,232 @@
+
 	switch t {
 	case wire.TBool:
-		b, off, err := br.readByte(off)
-		if err != nil {
-			return wire.Value{}, off, err
-		}
-
-		if b != 0 && b != 1 {
-			return wire.Value{}, off, decodeErrorf(
-				"invalid value %q for bool field", b,
-			)
-		}
-
-		return wire.NewValueBool(b == 1), off, nil
+		b, err := r.sr.ReadBool()
+		return wire.NewValueBool(b), r.or.offset, err
 
 	case wire.TI8:
-		b, off, err := br.readByte(off)
-		return wire.NewValueI8(int8(b)), off, err
+		b, err := r.sr.ReadInt8()
+		return wire.NewValueI8(int8(b)), r.or.offset, err
 
 	case wire.TDouble:
-		value, off, err := br.readInt64(off)
-		d := math.Float64frombits(uint64(value))
-		return wire.NewValueDouble(d), off, err
+		value, err := r.sr.ReadDouble()
+		return wire.NewValueDouble(value), r.or.offset, err
 
 	case wire.TI16:
-		n, off, err := br.readInt16(off)
-		return wire.NewValueI16(n), off, err
+		n, err := r.sr.ReadInt16()
+		return wire.NewValueI16(n), r.or.offset, err
 
 	case wire.TI32:
-		n, off, err := br.readInt32(off)
-		return wire.NewValueI32(n), off, err
+		n, err := r.sr.ReadInt32()
+		return wire.NewValueI32(n), r.or.offset, err
 
 	case wire.TI64:
-		n, off, err := br.readInt64(off)
-		return wire.NewValueI64(n), off, err
+		n, err := r.sr.ReadInt64()
+		return wire.NewValueI64(n), r.or.offset, err
 
 	case wire.TBinary:
-		v, off, err := br.readBytes(off)
-		return wire.NewValueBinary(v), off, err
+		v, err := r.sr.ReadBinary()
+		return wire.NewValueBinary(v), r.or.offset, err
 
 	case wire.TStruct:
-		s, off, err := br.readStruct(off)
-		return wire.NewValueStruct(s), off, err
+		s, err := r.readStructStream()
+		return wire.NewValueStruct(s), r.or.offset, err
 
 	case wire.TMap:
-		m, off, err := br.readMap(off)
-		return wire.NewValueMap(m), off, err
+		m, err := r.readMapStream()
+		return wire.NewValueMap(m), r.or.offset, err
 
 	case wire.TSet:
-		s, off, err := br.readSet(off)
-		return wire.NewValueSet(s), off, err
+		s, err := r.readSetStream()
+		return wire.NewValueSet(s), r.or.offset, err
 
 	case wire.TList:
-		l, off, err := br.readList(off)
-		return wire.NewValueList(l), off, err
+		l, err := r.readListStream()
+		return wire.NewValueList(l), r.or.offset, err
 
 	default:
-		return wire.Value{}, off, decodeErrorf("unknown ttype %v", t)
+		return wire.Value{}, r.or.offset, decodeErrorf("unknown ttype %v", t)
 	}
 }
+
+// Reader implements a parser for the Thrift Binary Protocol based on an
+// io.ReaderAt.
+type Reader struct {
+	reader io.ReaderAt
+}
+
+// NewReader builds a new Reader based on the given io.ReaderAt.
+func NewReader(r io.ReaderAt) Reader {
+	return Reader{reader: r}
+}
+
+// ReadValue reads a value off the given type off the wire starting at the
+// given offset.
+//
+// Returns the Value, the new offset, and an error if there was a decode error.
+func (br *Reader) ReadValue(t wire.Type, off int64) (wire.Value, int64, error) {
+	realReader := newReader(br.reader)
+	return realReader.ReadValue(t, off)
```

Comment:
nit: naming: realReader vs just reader? Or inline the variable?
```suggestion
	return newReader(br.reader).ReadValue(t, off)
```

Reply:
Inlined.